### PR TITLE
[wip] Tail recursion elimination

### PR DESCRIPTION
**Pull request by RealyUniqueName**
I tried to tailor it into the typing process to avoid creation of a separate filter, but I failed to find a way to do it without a finalized list of tail recursive function arguments at the moment of typing `TReturn`, which is not available at that moment. So I ended up with an additional expression filter behind `-D analyzer-optimize`.

TODO list:
 - [x] local named functions
 - [x] abstract member methods
 - [x] handle local vars captured in closures inside of tail recursive functions
 - [x] deal with recursive returns inside user-defined loops
 - [x] tail recursion for `Void` functions
 - [x] handle overriding

### CODE REVIEW

**Comment by Simn** [comment]
This doesn't seem to actually check for the "tail" part of tail-recursion, i.e. it picks up any self-recursive call which isn't correct. The `transform` function has to follow the control flow and analyze all leaves.

As a consequence, it is then not necessary to generate `continue` expressions because we must be at the end of the loop-body by definition.

**Comment by RealyUniqueName** [comment]
Why isn't it correct to eliminate any recursive return?
```haxe
function rec() {
    if(Std.random(2) == 0) return rec();
    doSomething();
    return 123;
}
```
into
```haxe
function rec() {
    while(true) {
        if(Std.random(2) == 0) {
            continue;
        }
        doSomething();
        return 123;
    }
}
```

**Comment by Simn** [comment]
Right, I suppose any `return` expression is technically in a tail-position. That's interesting, I never thought about TCO in that context.

One restriction we'll have to put in place is that we cannot do this if the recursive call is inside a loop. In that case the `continue` will not go to the right place.

**Comment by nadako** [comment]
> One restriction we'll have to put in place is that we cannot do this if the recursive call is inside a loop. In that case the `continue` will not go to the right place.

Can we technically generate some boolean flags to work that around? BTW Ideally we'd use `goto` on the targets that support it.

**Comment by RealyUniqueName** [comment]
Yes, we can. But I thought you want to complicate things later :)

**Comment by nadako** [comment]
Sure, just wondering if I'm missing something.

File: src-json/meta.json [hunk 1]
```diff
@@ -1101,6 +1101,11 @@
 		"platforms": ["java"],
 		"targets": ["TClass"]
 	},
+	{
+		"name": "TailRecursion",
+		"metadata": ":tailRecursion",
+		"doc": "Internally used for tail recursion elimination."
```
Comment (Gama11) [inline src-json/meta.json]: Missing `"internal": true`?

**Comment by Simn** [comment]
Does this respect overriding? A self-recursive call on `this` in a base class field doesn't necessarily call the same field at run-time.

**Comment by RealyUniqueName** [comment]
I've forgot about overriding. Will handle that.

**Comment by RealyUniqueName** [comment]
Done.

**Review by Simn — APPROVED** [review]
This requires some extensive testing, but the best way to achieve that is to just merge it and see what happens. The change overall is very isolated, so let's roll with it.

Let's also keep an eye on https://benchs.haxe.org/

**Pull request merged**